editor/src/components/editor/store/collaboration-state.tsx

```diff
@@ -0,0 +1,49 @@
+import React from 'react'
+import type { EditorDispatch } from '../action-types'
+import { releaseControlOverProject, snatchControlOverProject } from './collaborative-editing'
+import { Substores, useEditorState } from './store-hook'
+import { updateProjectServerState } from '../actions/action-creators'
+
+interface CollaborationStateUpdaterProps {
+  projectId: string | null
+  dispatch: EditorDispatch
+}
+
+export const CollaborationStateUpdater = React.memo(
+  (props: React.PropsWithChildren<CollaborationStateUpdaterProps>) => {
+    const { projectId, dispatch, children } = props
+    const isMyProject = useEditorState(
+      Substores.projectServerState,
+      (store) => store.projectServerState.isMyProject,
+      'CollaborationStateUpdater isMyProject',
+    )
+    React.useEffect(() => {
+      function handleVisibilityChange(): void {
+        if (projectId != null) {
+          // If the document is hidden, that means the editor is in the background
+          // or minimised, so release control. Otherwise if it has become unhidden
+          // then snatch control of the project.
+          if (document.hidden) {
+            void releaseControlOverProject(projectId).then(() => {
+              dispatch([updateProjectServerState({ currentlyHolderOfTheBaton: false })])
+            })
+          } else {
+            // Only attempt to do any kind of snatching of control if the project is "mine".
```

Comment:
Does it make sense to wrap the entire thing in `if (isMyProject === 'yes')` (including the releasing part) rather than just the snatching part?

Reply:
There's been talk that we might be able to hand over control to another user, in which case we want the release control to apply in that case. It's just that currently it's invalid to claim control if you're not the owner. Specifically it will throw an exception.